### PR TITLE
Skip serialization for json string 

### PR DESCRIPTION
**Pull request by YouCanKeepSilence**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/validation.js [hunk 1]
```diff
@@ -122,10 +122,24 @@ function wrapValidationError (result, dataVar) {
...
+  } catch (e) {
+    return false
+  }
+}
+
 function serialize (context, data, statusCode) {
   var responseSchemaDef = context[responseSchema]
   if (!responseSchemaDef) {
-    return JSON.stringify(data)
+    if (typeof data === 'string' && _isJsonString(data)) {
```
Comment (RafaelGSS) [inline lib/validation.js]: I think that it can affect the performance.... if we receive a parameter to explicit it?

File: lib/validation.js [hunk 2]
```diff
@@ -122,10 +122,24 @@ function wrapValidationError (result, dataVar) {
   return error
 }
 
+// no idea about performance
```
Comment (jsumners) [inline lib/validation.js]: The "idea" would be _slow_.

File: lib/validation.js (hunk 1 above)
Comment (YouCanKeepSilence) [reply lib/validation.js]: Yes, it'll be more efficient solution, but i don't know how to pass param here...

Comment (YouCanKeepSilence) [reply lib/validation.js]: I'll try to pass it to context

File: lib/validation.js (hunk 2 above)
Comment (mcollina) [reply lib/validation.js]: This will be slow unfortunately. We need a different solution for this.

**Comment by YouCanKeepSilence** [comment]
I'll try to add this as explicit param.

File: lib/reply.js [hunk 3]
```diff
@@ -195,6 +197,12 @@ Reply.prototype.headers = function (headers) {
   return this
 }
 
+Reply.prototype.disableStringify = function (flag) {
```
Comment (jsumners) [inline lib/reply.js]:
```suggestion
Reply.prototype.skipSerialization = function () {
```

And simply set the trackers to `true` such that `reply.skipSerialization().send('{"foo":"bar"}')` works.

Comment (YouCanKeepSilence) [reply lib/reply.js]: Sounds good, i'll change this.

File: docs/Reply.md [hunk 4]
```diff
@@ -334,6 +335,19 @@ fastify.get('/teapot', async function (request, reply) => {
 
 If you want to know more please review [Routes#async-await](https://github.com/fastify/fastify/blob/master/docs/Routes.md#async-await).
 
+<a name="skipserialization"></a>
+### .skipSerialization()
+
+Disable calling stringify method for already serialized payload.
+You should use it when your payload already serialized to string and conforms to yours selected Content-Type header value
```
Comment (jsumners) [inline docs/Reply.md]:
```suggestion
This is useful when the response payload is already serialized according to the response `Content-Type` header.
```

File: docs/Reply.md [hunk 5]
```diff
@@ -334,6 +335,19 @@ fastify.get('/teapot', async function (request, reply) => {
 
 If you want to know more please review [Routes#async-await](https://github.com/fastify/fastify/blob/master/docs/Routes.md#async-await).
 
+<a name="skipserialization"></a>
+### .skipSerialization()
+
+Disable calling stringify method for already serialized payload.
```
Comment (jsumners) [inline docs/Reply.md]:
```suggestion
Disable automatic serialization for the current reply.
```

File: test/internals/reply.test.js [hunk 6]
```diff
@@ -544,6 +544,30 @@ test('plain string with content type application/json should be serialized as js
   })
 })
 
+test('string with content type application/json and skipSerialization() called should NOT be serialized as json', t => {
```
Comment (RafaelGSS) [inline test/internals/reply.test.js]: Could add another test to `JSON.stringfy({ test: true })`? Just to prevent something

Comment (YouCanKeepSilence) [reply test/internals/reply.test.js]: What you mean? Simple send serialized via JSON.stringify data with skipSerialization()?

**Comment by YouCanKeepSilence** [comment]
Don't know the reason why checks fails? Can't find the error log. At local machine 'test', 'test:ci', 'benchmark' scripts run correctly

File: test/internals/reply.test.js (hunk 6 above)
Comment (RafaelGSS) [reply test/internals/reply.test.js]: Yes

**Comment by RafaelGSS** [comment]
> Don't know the reason why checks fails? Can't find the error log. At local machine 'test', 'test:ci', 'benchmark' scripts run correctly

I think that is a little bug inside our CI, timeout or something like it. Don't worry.

**Comment by YouCanKeepSilence** [comment]
> > Don't know the reason why checks fails? Can't find the error log. At local machine 'test', 'test:ci', 'benchmark' scripts run correctly
> 
> I think that is a little bug inside our CI, timeout or something like it. Don't worry.

So i should just wait?

File: lib/reply.js [hunk 7]
```diff
@@ -195,6 +197,12 @@ Reply.prototype.headers = function (headers) {
   return this
 }
 
+Reply.prototype.skipSerialization = function () {
+  this[kReplyskipSerialization] = true
+  this.context[kReplyskipSerialization] = true
```
Comment (mcollina) [inline lib/reply.js]: Unfortunately this won't work. The context is shared between multiple execution of the handler and multiple reply objects. Can you remove it? The alternative is to move this to the Context itself, and set this option during the route creation.

Comment (YouCanKeepSilence) [reply lib/reply.js]: You told about this.context? What about just this? Will it work If i pass from this[kReplyskipSerialization] to serialize() method as optional params?

Comment (mcollina) [reply lib/reply.js]: I would actually change the serialize function to receive the Reply object instead.

Comment (YouCanKeepSilence) [reply lib/reply.js]: Changed it to pass this param via `this` of reply object. Added test to check that `skipSerialization` call doesn't affect next responses.

**Comment by YouCanKeepSilence** [comment]
So what about checks? What is incorrect?

**Comment by mcollina** [comment]
It's just flaky CI

**Comment by YouCanKeepSilence** [comment]
> It's just flaky CI

So it'll be approved later? Or you just merge it with ignore this checks?

**Comment by mcollina** [comment]
Let's wait another review (cc @jsumners), then it can land.

File: test/internals/reply.test.js [hunk 8]
```diff
@@ -14,6 +14,8 @@ const {
   kReplyIsError
 } = require('../../lib/symbols')
 
+t.setTimeout(0)
```
Comment (jsumners) [inline test/internals/reply.test.js]: Everything looks good, but why this line?

Comment (YouCanKeepSilence) [reply test/internals/reply.test.js]: Oh, sorry, I forgot remove it after debugging

**Comment by YouCanKeepSilence** [comment]
As mentioned, there is no fast method to check, that string contains valid json.

**Comment by delvedor** [comment]
Again, maybe I'm missing something. What I would do, is to check internally the `content-type` header, and if it is `application/json`, and the payload is already a string, then we should skip the serialization.
We don't care if the string is a valid json or not, as is the responsibility of the user.

**Comment by YouCanKeepSilence** [comment]
In your tests already exists test that checks that simple string with `content-type` must be serialized as json. So this change will break some existing functionality. What you think about it? Should I change test?

**Comment by Eomm** [comment]
If the user send a json reply with the string "hi" he would get '"hi"' right?

And if we skip without a new parameter, or the check of the header, it would a possible breaking change. 

I don't have a pc to try till monday, could someone confirm the first assumption?

**Comment by YouCanKeepSilence** [comment]
> If the user send a json reply with the string "hi" he would get '"hi"' right?
> 
Yes, right now in test exists this case
> And if we skip without a new parameter, or the check of the header, it would a possible breaking change.
>
Right

**Comment by delvedor** [comment]
> In your tests already exists test that checks that simple string with content-type must be serialized as json.

Can you provide a link to the test? Thanks!

---

I've just tried the following code:
```js
'use strict'

const fastify = require('fastify')()

fastify.get('/', (req, reply) => {
  reply.send('hello world')
})

fastify.listen(3000, console.log)
```
And this is the result:
```sh
▶ curl -v http://localhost:3000
* Rebuilt URL to: http://localhost:3000/
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 3000 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< content-type: text/plain; charset=utf-8
< content-length: 11
< Date: Sun, 17 Nov 2019 13:37:35 GMT
< Connection: keep-alive
<
* Connection #0 to host localhost left intact
hello world% 
```
Which is correct. If I add the `application/json` content type instead:
```js
'use strict'

const fastify = require('./')()

fastify.get('/', (req, reply) => {
  reply.type('application/json').send('hello world')
})

fastify.listen(3000, console.log)
```
The result is
```sh
▶ curl -v http://localhost:3000
* Rebuilt URL to: http://localhost:3000/
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 3000 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< content-type: application/json; charset=utf-8
< content-length: 13
< Date: Sun, 17 Nov 2019 13:39:04 GMT
< Connection: keep-alive
<
* Connection #0 to host localhost left intact
"hello world"% 
```

Which we all agree is not correct.
Internally if the content type is `application/json` and the payload is already a string, we should not serialize it again.

The following lines cause the issue:
https://github.com/fastify/fastify/blob/480dd934de1ed5b55ffca531a6e88539b20b5bd7/lib/reply.js#L139-L146
We can easily fix this if we add a check of the type of the payload.

**Comment by YouCanKeepSilence** [comment]
> > In your tests already exists test that checks that simple string with content-type must be serialized as json.
> 
> Can you provide a link to the test? Thanks!
> 
https://github.com/fastify/fastify/blob/eafafe24e6160d49effe04b43a3ef3d65afc8697/test/internals/reply.test.js#L523

**Comment by delvedor** [comment]
Thank you @YouCanKeepSilence.
@fastify/core I don't know why we have that test, probably we wanted to improve the code coverage (it was two years ago!).
I think this should be considered a bug and we should fix it, it might cause a breaking change, I can't tell if people are relying on this.
I propose to fix it and revert the commit if we see that it causes some problems in our community. If so, we'll release the fix in the next major.
What do you think?

**Comment by YouCanKeepSilence** [comment]
> Thank you @YouCanKeepSilence.
> @fastify/core I don't know why we have that test, probably we wanted to improve the code coverage (it was two years ago!).
> I think this should be considered a bug and we should fix it, it might cause a breaking change, I can't tell if people are relying on this.
> I propose to fix it and revert the commit if we see that it causes some problems in our community. If so, we'll release the fix in the next major.
> What do you think?

So, how i should fix this? Add explicit flag or just check typeof payload?

**Comment by Eomm** [comment]
> What do you think?

I think this is a bug but, I think it would be better to release in v2 with a flag, and in v3 without the flag skipping the serialization by default.

I agree that releasing it in v2 like a patch would not break because so much:
- devs use http-client that parse the response
- usually, the string data (from redis) are parsed to be read, processed and then sent to the client

but these are only my guesses and the opener of this PR would have had a break change 😅

**Comment by YouCanKeepSilence** [comment]
> What do you think?

I agree with @Eomm, so as i think more acceptable solution now is add explicit flag, and later change behaviour of this functionality. So if we talking about explicit flag i made it in current PR:)

**Comment by YouCanKeepSilence** [comment]
So, what about merge this fix? Or you need changes?

**Comment by mcollina** [comment]
@delvedor and myself had a quick chat about this, and we came to the conclusion that this is a bug and that test is wrong.
This was added to fix https://github.com/fastify/fastify/issues/540 and https://github.com/fastify/fastify/pull/541 (a long time ago). None had the intention of implementing the behavior that this PR is fixing.
I think changing that serialization is skipped when the content-type is specified is what we intended.
Therefore, we propose to treat that as a bugfix and release it in v2.

**Comment by YouCanKeepSilence** [comment]
> @delvedor and myself had a quick chat about this, and we came to the conclusion that this is a bug and that test is wrong.
> This was added to fix #540 and #541 (a long time ago). None had the intention of implementing the behavior that this PR is fixing.
> I think changing that serialization is skipped when the content-type is specified is what we intended.
> Therefore, we propose to treat that as a bugfix and release it in v2.

So you'll make it by yourself and my PR is not needed?(

**Comment by delvedor** [comment]
@YouCanKeepSilence you can update this pr and follow the suggestions I wrote in https://github.com/fastify/fastify/pull/1937#issuecomment-554747295.
Otherwise, you can open a new pr and reference this one :)

**Comment by YouCanKeepSilence** [comment]
I still don't know why checks fails, all test and benchs were ok.

**Comment by adrai** [comment]
Is it intended that with this fix also such a response now comes different with v2.12?

example: `reply.send('5d4a5dcf-f265-499d-94fe-644144c323dd')`
response with v2.12: `5d4a5dcf-f265-499d-94fe-644144c323dd`
response with v2.11: `"5d4a5dcf-f265-499d-94fe-644144c323dd"`

in v2.11 it was a json and in v1.12 it is now a string

**Comment by jsumners** [comment]
That definitely seems like an out of spec regression.

**Comment by mcollina** [comment]
@jsumners it’s exacty what this PR does. We might consider reverting it in v2 and applying it only in v3.

**Comment by jsumners** [comment]
@mcollina I thought the PR made it an option to skip serialization of a string.

**Invalid JSON:**
```json
foo
```

**Valid JSON:**
```json
"foo"
```

**Comment by adrai** [comment]
for me there is no problem... have a workaround...
imo this PR is ok for v2, but maybe there should also be a fix to handle a string correctly in case the content type is application/json

**Pull request merged**